tik/lib/CartographerKernel.cpp

```diff
@@ -192,95 +220,57 @@ namespace TraceAtlas::tik
-            uint64_t i;
-            for (i = 0; i < KernelImports.size(); i++)
-            {
-                auto *a = cast<Argument>(KernelFunction->arg_begin() + 1 + i);
-                a->setName("i" + to_string(i));
-                VMap[IDToValue[KernelImports[i]]] = a;
-                ArgumentMap[a] = KernelImports[i];
-            }
-            uint64_t j;
-            for (j = 0; j < KernelExports.size(); j++)
-            {
-                auto *a = cast<Argument>(KernelFunction->arg_begin() + 1 + i + j);
-                a->setName("e" + to_string(j));
-                ArgumentMap[a] = KernelExports[j];
-            }
+            set<Function *> embeddedKernels;
+            ConstructFunctionSignature(blocks, embeddedKernels, VMap, name);
 
             //create the artificial blocks
             Init = BasicBlock::Create(TikModule->getContext(), "Init", KernelFunction);
             Exit = BasicBlock::Create(TikModule->getContext(), "Exit", KernelFunction);
             Exception = BasicBlock::Create(TikModule->getContext(), "Exception", KernelFunction);
 
-            //copy the appropriate blocks
             BuildKernelFromBlocks(VMap, blocks);
 
-            Remap(VMap); //we need to remap before inlining
+            BuildInit(VMap);
+
+            // we need to remap before inlining
+            Remap(VMap);
 
             InlineFunctionsFromBlocks(blockSet);
 
             CopyGlobals(VMap);
 
-            //remap and repipe
             Remap(VMap);
 
-            // replace external function calls with tik declarations
-            for (auto &bi : *(KernelFunction))
+            // patch work here. Sometimes when inlining, llvm will inject the block terminator before an export store
+            for (auto &fi : *KernelFunction)
             {
-                for (auto inst = bi.begin(); inst != bi.end(); inst++)
+                if (auto st = dyn_cast<StoreInst>(prev(fi.end())))
                 {
-                    if (auto callBase = dyn_cast<CallBase>(inst))
+                    // have to find the true terminator
+                    for (auto &bi : fi)
                     {
-                        Function *f = callBase->getCalledFunction();
-                        if (f == nullptr)
+                        if (bi.isTerminator())
                         {
-                            throw AtlasException("Null function call (indirect call)");
+                            bi.moveAfter(st);
                         }
-
-                        auto *funcDec = cast<Function>(TikModule->getOrInsertFunction(callBase->getCalledFunction()->getName(), callBase->getCalledFunction()->getFunctionType()).getCallee());
-                        funcDec->setAttributes(callBase->getCalledFunction()->getAttributes());
-                        callBase->setCalledFunction(funcDec);
                     }
                 }
             }
```

Comment:
I'm not sure what this code is doing. It used to be a catch for indirect function calls, but that appears to have been removed and replaced with an inlining check?

Reply:
Ah, so this is misleading. These two things are actually not replacements of each other, even though the diffs appear to be so.

The check for indirect function calls has been moved to FindScopedStructures() at the top of this file. Therefore function pointers don't need to be checked here.

Now this check is on instructions created by me that have accidentally found themselves as the block terminator (my case test is for Halide/fft program where a store instruction to an export is for some reason a block terminator).